lib/ecto/repo/queryable.ex

```diff
@@ -74,6 +74,37 @@ defmodule Ecto.Repo.Queryable do
     one!(name, query_for_get_by(queryable, clauses), opts)
   end
 
+  def reload(name, structs, opts) when is_list(structs) do
+    all(name, query_for_reload(structs, opts), opts)
+  end
+
+  def reload(name, struct, opts) do
+    one(name, query_for_reload(struct, opts), opts)
+  end
+
+  def reload!(name, structs, opts) when is_list(structs) do
+    query = query_for_reload(structs, opts)
+    result = all(name, query, opts)
+    result_count = Enum.count(result)
+    expected_count = Enum.count(structs)
+
+    if result_count < expected_count do
+      raise Ecto.TooFewResultsError, queryable: query, min: expected_count, count: result_count
+    end
+
+    result
```

Comment:
Unfortunately there is no guarantee `result` will be in the same order as `structs`. So you need to do something like this:

    for struct <- structs do
      struct_pk = Map.fetch!(struct, pk)
      Enum.find(result, &Map.fetch!(&1, pk) == struct_pk) || raise "could not find"
    end

This means you don't need the Enum.count nor the exception above, as the above will guarantee all entries have a match.

Reply:
If we guarantee same ordering in `reload!/2`, should we also do the same for `reload/2`?

Reply:
Yes, it should be for both!